### PR TITLE
feat: collection step for seedlot registration

### PR DESCRIPTION
**Pull request by iatdaitan**
# Description

Please provide a summary of the change and the issue fixed. Please include relevant context. List dependency changes.

Fixes [#549](https://apps.nrs.gov.bc.ca/int/jira/browse/FSADT2-549)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Tested locally using Windows (WSL 2) and Chrome-based browser.

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Frontend](https://nr-spar-webapp-79-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar-webapp/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Frontend](https://nr-spar-webapp-79-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar-webapp/actions/workflows/merge-main.yml)

### CODE REVIEW

**Comment by SLDonnelly** [comment]
Hi there, I was just looking at the screen and noticed a potential bug. The second field highlighted in yellow in the screenshot should be for the number of bags rather than the collection agency

image

**Comment by SLDonnelly** [comment]
Hello! In the collection form, could we please remove the hard coding for location code please and allow this to be a 2 digit field that is entered by the user? 
image
image

**Comment by iatdaitan** [comment]
> Hello! In the collection form, could we please remove the hard coding for location code please and allow this to be a 2 digit field that is entered by the user? image image

Yes, sure. This will be done once I finish the integration with the mock.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 1]
```diff
@@ -0,0 +1,320 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (mgaseta) [inline frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:1]: Remove this eslint-disable, I think this was left here on a first version of the changes, it's not necessary anymore.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+              ref={volumePerContainersInputRef}
+              value={state.volumePerContainers}
+              label={fieldsConfig.volumePerContainers.labelText}
+              invalidText={fieldsConfig.volumePerContainers.invalidText}
+              onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                handleFormInput('volumePerContainers', e.target.value);
+              }}
+              hideSteppers
+              disableWheel
+            />
```
Comment (mgaseta) [inline frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:263]: On figma, the default value of this field seems to be 1, not sure if that's correct, maybe align with the UX team?
image

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+              value={state.volumeOfCones}
+              label={fieldsConfig.volumeOfCones.labelText}
+              invalidText={fieldsConfig.volumeOfCones.invalidText}
+              helperText={fieldsConfig.volumeOfCones.helperText}
+              onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                handleFormInput('volumeOfCones', e.target.value);
+              }}
+              hideSteppers
+              disableWheel
+            />
```
Comment (mgaseta) [inline frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:280]: On figma, the default value of this field seems to be 1, not sure if that's correct, maybe align with the UX team? Also on figma, the value of this field should be calculated by the system.
image

File: frontend/src/types/CollectionInformation.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+type CollectionInformation = {
+    seedlotNumber: string;
+    applicant: CollectorAgency;
+    startDate?: string;
+    endDate?: string;
+    numberOfContainers?: string;
+    volumePerContainer?: string;
+    volumeOfCones?: string;
+    collectionMethods?: CollectionMethods;
+    comments?: string;
```
Comment (mgaseta) [inline frontend/src/types/CollectionInformation.ts:13]: Fix indentation here.

**Review by mgaseta — CHANGES_REQUESTED** [review]
Nice job! Loved how the code design changes we discussed are already implemented ❤!
Added a few markups related to some missing functionalities and other minor changes.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx (hunk 1 above)
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx]: Fixed. Please review it again.

File: frontend/src/types/CollectorAgency.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+type CollectorAgency = {
+    name?: string;
+    locationCode?: string;
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/types/CollectorAgency.ts]: Fixed. Please review it again.

File: frontend/src/types/CollectionInformation.ts (hunk 4 above)
Comment (iatdaitan) [reply frontend/src/types/CollectionInformation.ts]: Fixed. Please review it again.

File: frontend/src/types/CollectorAgency.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+}
+
+export default CollectorAgency;
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/types/CollectorAgency.ts]: Fixed. Please review it again.

File: frontend/src/types/CollectionInformation.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+}
+
+export default CollectionInformation;
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/types/CollectionInformation.ts]: Fixed. Please review it again.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/styles.scss [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+.collector-information-dates {
+  margin-top: 2.5rem;
+
+  .#{vars.$bcgov-prefix}--date-picker--single {
+		width: 100%;
+		
+		.#{vars.$bcgov-prefix}--date-picker-container,  .#{vars.$bcgov-prefix}--date-picker__input {
+			width: 100%;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/styles.scss]: Fixed. Please review it again.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx (hunk 2 above)
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:262]: Fixed. Please review it again.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+              ref={numberOfContainersInputRef}
+              value={state.numberOfContainers}
+              label={fieldsConfig.numberOfContainers.labelText}
+              invalidText={fieldsConfig.numberOfContainers.invalidText}
+              onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                handleFormInput('numberOfContainers', e.target.value);
+              }}
+              hideSteppers
+              disableWheel
+            />
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:248]: Fixed. Please review it again.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx (hunk 3 above)
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:279]: I am checking with the team about the calculation and the warning message. I will fix it soon.

Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx]: @mgaseta Could you please review it again?

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+            <fieldset>
+              <Subtitle text="Collection methods (Select at least one method used to collect the cones)" />
+              <Checkbox labelText="Aerial raking" />
+              <Checkbox labelText="Aerial clipping or topping" />
+              <Checkbox labelText="Felled trees" />
+              <Checkbox labelText="Climbing" />
+              <Checkbox labelText="Squirrel cache" />
+              <Checkbox labelText="Ground, Ladder and or Hydraulic lift" />
+              <Checkbox labelText="Squirrel harvesting or dropping" />
+              <Checkbox labelText="Other" />
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx]: @mgaseta Could you please review it again?

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 11]
```diff
@@ -107,20 +121,50 @@ const CollectionStep = (
...
+    );
+  };
+
+  const collectionVolumeInformationHandler = <K extends keyof CollectionForm>(
+    name: K,
+    value: CollectionForm[K]
+  ) => {
+    const numberOfContainers = +refControl.current[fieldsConfig.numberOfContainers.name].value;
+    const volumePerContainers = +refControl.current[fieldsConfig.volumePerContainers.name].value;
+    const volumeOfCones = +refControl.current[fieldsConfig.volumeOfCones.name].value;
```
Comment (mgaseta) [inline frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:141]: Can't this information come directly from the `state` prop?

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 12]
```diff
@@ -136,29 +180,36 @@ const CollectionStep = (
...
-              }
+              checked={isChecked}
+              onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                collectorAgencyIsChecked(e);
+              }}
             />
           </Column>
         </Row>
         <Row>
           <Column sm={12} md={4} lg={6}>
```
Comment (mgaseta) [inline frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:195]: Didn't noticed this before, sorry... But the column's sizes are a bit different with carbon, please follow this table: 
image
Leaving only this markup, but please fix it on the other columns as well.

**Review by mgaseta — CHANGES_REQUESTED** [review]
Nice job ✔🥇! Added only a few more points.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx (hunk 11 above)
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:141]: Tried it before but had a conflict with the state.

File: frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx [hunk 13]
```diff
@@ -284,27 +363,122 @@ const CollectionStep = (
...
+                  placeholder={fieldsConfig.collectionMethod.placeholder}
+                  helperText={fieldsConfig.collectionMethod.helperText}
+                  value={state.collectionMethodName}
+                  onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+                    handleFormInput(
+                      fieldsConfig.collectionMethod.name as keyof CollectionForm,
+                      e.target.value
+                    );
+                  }}
+                />
```
[the remark this answers is not in the document]
Comment (iatdaitan) [reply frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx]: Please review it again.

**Review by mgaseta — APPROVED** [review]
LGTM! 🐱‍🏍🚀🎉

**Review by RMCampos — APPROVED** [review]
Good job! 🥳

**Pull request merged**